cranelift/codegen/src/isa/riscv64/inst.isle

```diff
@@ -2243,13 +2299,155 @@
+
+
+;; This emits just the comparision instructions and assumes that
+;; the arguments are already extended.
+;;
+;; We only have actual lowerings for Equal/NotEqual/SignedLessThan/UnsignedLessThan.
+;; Everything else just recurses into one of those cases.
+(decl gen_icmp_inner (IntCC ValueRegs ValueRegs Type) Reg)
+
+;; We only implement LessThan, for GreaterThan we just reverse the arguments
+(rule 3 (gen_icmp_inner (intcc_greater_than cc) x y ty)
+  (gen_icmp_inner (intcc_reverse cc) y x ty))
+
+;; For these rules, we can just use the normal rules and then invert the result.
+;; i.e. `x <= y` is the same as `!(x > y)`.
+(rule 2 (gen_icmp_inner (intcc_sle_or_ule cc) x y ty)
+  (let ((res Reg (gen_icmp_inner (intcc_inverse cc) x y ty)))
+    (rv_xori res (imm12_const 1))))
+
+
+;; For `*LessThan` we have a dedicated instruction `slt`/`sltu`.
+(rule (gen_icmp_inner (IntCC.SignedLessThan) x y (fits_in_64 ty))
+  (rv_slt x y))
+(rule (gen_icmp_inner (IntCC.UnsignedLessThan) x y (fits_in_64 ty))
+  (rv_sltu x y))
+
+;; We have a equivalent pseudo instruction for I128 less than.
+(rule 1 (gen_icmp_inner (IntCC.SignedLessThan) x y $I128)
+  (emit_i128_less_than x y (Signedness.Signed)))
+(rule 1 (gen_icmp_inner (IntCC.UnsignedLessThan) x y $I128)
+  (emit_i128_less_than x y (Signedness.Unsigned)))
+
+
+;; Compare both registers using xor, and set the result using the dedicated
+;; `seqz`/`snez` instructions.
+(rule (gen_icmp_inner (IntCC.Equal) x y (fits_in_64 ty))
+  (rv_seqz (rv_xor x y)))
+(rule (gen_icmp_inner (IntCC.NotEqual) x y (fits_in_64 ty))
+  (rv_snez (rv_xor x y)))
+
+;; In the I128 case we just `xor` everything and check if its zero at the end.
+(rule 1 (gen_icmp_inner (IntCC.Equal) x y $I128)
+  (let ((top_eq Reg (rv_xor (value_regs_get x 0) (value_regs_get y 0)))
+        (bottom_eq Reg (rv_xor (value_regs_get x 1) (value_regs_get y 1)))
+        (res Reg (rv_or top_eq bottom_eq)))
+    (rv_seqz res)))
+(rule 1 (gen_icmp_inner (IntCC.NotEqual) x y $I128)
+  (let ((top_eq Reg (rv_xor (value_regs_get x 0) (value_regs_get y 0)))
+        (bottom_eq Reg (rv_xor (value_regs_get x 1) (value_regs_get y 1)))
+        (res Reg (rv_or top_eq bottom_eq)))
+    (rv_snez res)))
+
+
+
+
+;; For some icmp's we can optimize the instruction sequence if the RHS is a constant.
+;;
+;; TODO: Currently we only have rules for <=64bit types. We can add more rules for I128's
+(decl gen_icmp_imm (IntCC ValueRegs Imm12 Type) Reg)
+
+;; This deals with the case where we have an iconst.i8 that is larger than 127.
+;; This is allowed by clif semantics. However when using this constant as a signed value
+;; we need to view it as a negative value, otherwise the comparison will be wrong.
+;;
+;; This can only happen with I8's, since it's the only type smaller than 12bits.
+(rule 6 (gen_icmp_imm cc x imm ty @ $I8)
+  (if-let cc (signed_cond_code cc))
+  (if-let $true (imm12_is_outside_signed_range imm ty))
+  (gen_icmp_imm cc x (neg_imm12 (imm12_and imm 0x7F)) ty))
```

Comment:
I'm not sure this is the best way of dealing with this. I'd ideally like to do something that would solve this for all users of `Imm12`, but I can't think of a way to do that without introducing a bunch of unnecessary details everywhere.


The testcase that raised this point was `runtests/icmp.imm` (cc: #3056):
```
; This test is also a regression test for aarch64.
; We were not correctly handling the fact that the rhs constant value
; overflows its type when viewed as a signed value, and thus encoding the wrong
; value into the resulting instruction.
function %overflow_rhs_const(i8) -> i8 {
block0(v0: i8):
    v1 = iconst.i8 192
    v2 = icmp sge v0, v1
    return v2
}
; run: %overflow_rhs_const(49) == 1
; run: %overflow_rhs_const(-65) == 0
```

Reply:
It seems like the basic primitive we want here is sign extension? I agree that special-casing i8 here doesn't feel right otherwise.

Reply:
I think the pattern we should always use is to either sign-extend or zero-extend constants according to how they're being used. We have `i64_sextend_imm64` and `u64_uextend_imm64` constructors for this. I guess you have `Imm12` at this point but maybe you should change that?

Reply:
Sorry for taking a while to get back to this! I've played around a bunch with this, I think I found something that I like along these lines.